src/dmd/cppmangle.d

```diff
@@ -101,8 +101,52 @@ bool isPrimaryDtor(const Dsymbol sym)
     return dtor == ad.primaryDtor;
 }
 
+/// Context used when processing pre-semantic AST
+private struct Context
+{
+    /// Template instance of the function being mangled
+    TemplateInstance ti;
+    /// Function declaration we're mangling
+    FuncDeclaration fd;
+    /// Current type / expression being processed (semantically analyzed)
+    RootObject res;
+
+    @disable ref Context opAssign (ref Context other);
+    @disable ref Context opAssign (Context other);
+
+    /**
+     * Helper function to track `res`
+     *
+     * Params:
+     *   next = Value to set `this.res` to
+     *          If `this.res` is `null`, the expression is not evalutated
+     *          This allow this code to be used even when no context is needed.
+     *
+     * Returns:
+     *   The previous state of this `Context` object
+     */
+    private Context set (lazy RootObject next)
+    {
+        auto r = this.res;
+        if (r !is null)
+            this.res = next;
+        return Context(this.ti, this.fd, r);
+    }
+
+    /**
+     * Reset the context to a previous one, making any adjustment necessary
+     */
+    private void reset (ref Context prev)
+    {
+        this.res = prev.res;
+    }
+}
+
 private final class CppMangleVisitor : Visitor
 {
+    /// Context used when processing pre-semantic AST
+    private Context ctx;
```

Comment:
Can we please avoid abbreviations like this.

Reply:
I wish you could convince Walter, because `ctx` is much more widespread than the various 2-letters type abbreviations which are so common in DMD. Anyway, I'll update with the full name.

Reply:
A project wide search for `ctx` gives back 25 results and for `context` it gives back 253 results. This is just a plain text search which includes comments and everything.